routines/analyze.go

```diff
@@ -0,0 +1,103 @@
+package routines
+
+import (
+	"log"
+	"time"
+
+	"github.com/csivitu/bl0b/ctftime"
+
+	"github.com/csivitu/bl0b/db"
+)
+
+func computeStatus(start time.Time, end time.Time) string {
+	t := time.Now().Unix()
+	startUnix := start.Unix()
+	endUnix := end.Unix()
+
+	if startUnix <= t && t < endUnix {
+		return "ongoing"
+	}
+
+	if t < startUnix && t < endUnix {
+		return "upcoming"
+	}
+
+	return "over"
+}
+
+func handleUpcoming(event *ctftime.Event, DB *db.Database) {
+	if event.Status == "upcoming" {
+		return
+	}
+
+	err := DB.ModifyEventStatus(event.ID, "upcoming")
+	if err != nil {
+		log.Println("Error while modifying event " + event.Title + " in handleUpcoming")
+	}
+}
+
+func handleOngoing(event *ctftime.Event, DB *db.Database) {
+	// If the status is already ongoing, return
+	if event.Status == "ongoing" {
+		return
+	}
+
+	err := DB.ModifyEventStatus(event.ID, "ongoing")
+	if err != nil {
+		log.Println("Error while modifying event " + event.Title + " in handleOngoing")
+		log.Println(err)
+	}
+}
+
+func handleOver(event *ctftime.Event, DB *db.Database) {
+	err := DB.DeleteEventByID(event.ID)
+	if err != nil {
+		log.Println("Error while deleting event " + event.Title)
+		log.Println(err)
+	}
+}
+
+func analyze() {
```

Comment:
Just as a side note, wouldn't it be possible to just do this in a single Update DB query (since SQL can do timestamp comparisions)?
Currently this does O(n) db calls, while I think it can be constant time.

Reply:
I structured this inside a for loop since I was planning to be able to run a `notify` function inside `handleOngoing()`, i.e, if the status is `ongoing` but in the DB it's set as `upcoming`, call a notify function to send DMs to all channels about the CTF.

If you do it in O(1) using SQL, the only way I can think of for getting notifications to work would be a third `goroutine` just for the purpose of notifying, which I don't think is necessary. That routine would also be stateful since it would have to remember the previous state of the `Status` field in the DB since it doesn't know when it changed.

Reply:
I see that makes sense